src/jobserver-posix.cc

```diff
@@ -0,0 +1,170 @@
+// Copyright 2024 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "jobserver.h"
+
+#include <fcntl.h>
+#include <unistd.h>
+
+#include <cassert>
+#include <cstring>
+
+#include "util.h"
+
+Jobserver::Jobserver() {
+  assert(!Enabled());
+
+  // Return early if no makeflags are passed in the environment.
+  char* makeflags = std::getenv("MAKEFLAGS");
+  if (makeflags == nullptr || strlen(makeflags) == 0)
+    return;
+
+  std::string::size_type flag_char_ = 0;
+  std::string flag_;
+  std::vector<std::string> flags_;
+
+  // Tokenize string to characters in flag_, then words in flags_.
+  while (flag_char_ < strlen(makeflags)) {
+    while (flag_char_ < strlen(makeflags) &&
+           !isblank(static_cast<unsigned char>(makeflags[flag_char_]))) {
+      flag_.push_back(static_cast<unsigned char>(makeflags[flag_char_]));
+      flag_char_++;
+    }
+
+    if (!flag_.empty())
+      flags_.push_back(flag_);
+
+    flag_.clear();
+    flag_char_++;
+  }
+
+  // --jobserver-auth=<val>
+  for (size_t n = 0; n < flags_.size(); n++)
+    if (flags_[n].find(AUTH_KEY) == 0)
+      flag_ = flags_[n].substr(strlen(AUTH_KEY));
+
+  // --jobserver-fds=<val>
```

Comment:
Note that this option is never going to work unless you can modify subprocess-posix.cc and subprocess-win32.cc to pass these extra file descriptors to all child processes. Which is a very non-trivial change. Do you have a commit planned for this? Otherwise, I would leave this out of the current PR, explaning why it's not implemented yet.

(Alternatively, support it in the parser function, but ensure that the client recognizes this is not implemented yet, and ignore it).

Reply:
> Note that this option is never going to work unless you can modify subprocess-posix.cc and subprocess-win32.cc to pass these extra file descriptors to all child processes. 

Are you sure about that? I don't see a difference in functionality between Make 4.1 and Make 4.2 for the jobserver. There was a lot of refactoring going on, but the commit that renames the flag is literally just a rename. I'll do some testing regarding this though.

> Do you have a commit planned for this? Otherwise, I would leave this out of the current PR, explaning why it's not implemented yet.

No, I don't have a commit for that. I imagine that any problem with the simple pipe with invalid FDs passed will fallback to non-parallel building.

Reply:
@digit-google funny enough, instead of finding a problem between Make 4.1 and Make 4.2, I found a problem between Make 4.2.1 and Make 4.3. This bug makes any process into a zombie if and only if they try to read a token from the token server and the token is not available AND there is no other subprocess that will finish so that a token is delivered back into the pool to be redirected to the zombie process so it can run. In my testing use case this makes `make -j 3` fail while `make -j` without a number works.

This is because the pipe used to be blocking on the write side of the server, the case described by POSIX as:

> If some process has the pipe open for writing and O_NONBLOCK is clear, read() will block the calling thread until some data is written or the pipe is closed by all processes that had the pipe open for writing.

One way we can adapt is to check the NONBLOCK flag and force non-parallel if it is cleared. However, most of the responsibility here should be to the user who should know the limitations of Make 4.2.1 and earlier.

bug:
https://savannah.gnu.org/bugs/?51159

fixed with commit "[SV 51159] Use a non-blocking read with pselect to avoid hangs.":
https://git.savannah.gnu.org/cgit/make.git/commit/?id=b552b05251980f693c729e251f93f5225b400714

Reply:
A little more context, the way the client is currently written, ninja tries to get a second token before returning the first one it just got, so this Make bug causes a ninja build to hang even if it is not multiple recursive levels deep.